network/src/peer/fuzzing.rs

```diff
@@ -0,0 +1,150 @@
+/// out to a buffer using our length-prefixed message codec.
+pub fn generate_corpus(gen: &mut ValueGenerator) -> Vec<u8> {
+    let network_msgs = gen.generate(vec(any::<NetworkMessage>(), 1..20));
+
+    let network_msgs = network_msgs
+        .into_iter()
+        .map(|msg| lcs::to_bytes(&msg).unwrap())
+        .collect::<Vec<_>>();
+
+    let (write_socket, mut read_socket) = MemorySocket::new_pair();
+    let mut writer = FramedWrite::new(IoCompat::new(write_socket), message_codec(MAX_FRAME_SIZE));
+
+    // Write the `NetworkMessage`s to a fake socket
+    let f_send = async move {
+        for network_msg in network_msgs {
+            writer.send(network_msg.into()).await.unwrap();
+        }
+    };
+    // Read the serialized `NetworkMessage`s from the fake socket
+    let f_recv = async move {
+        let mut buf = Vec::new();
+        read_socket.read_to_end(&mut buf).await.unwrap();
+        buf
+    };
+
+    let (_, buf) = block_on(future::join(f_send, f_recv));
+    buf
+}
+
+/// Fuzz the `Peer` actor's inbound message handling.
+///
+/// For each fuzzer iteration, we spin up a new `Peer` actor and pipe the raw
+/// fuzzer data into it. This mostly tests that the `Peer` inbound message handling
+/// doesn't panic or leak memory when reading, deserializing, and handling messages
+/// from remote peers.
+pub fn fuzz(data: &[u8]) {
+    // Use the basic single-threaded runtime, since our current tokio version has
+    // a chance to leak memory and/or thread handles when using the threaded
+    // runtime and sometimes blocks when trying to shutdown the runtime.
+    //
+    // https://github.com/tokio-rs/tokio/pull/2649
+    let mut rt = tokio::runtime::Builder::new()
+        .basic_scheduler()
+        .enable_all()
+        .build()
+        .unwrap();
+    let executor = rt.handle().clone();
+
+    // We want to choose a constant peer id for _our_ peer id, since we will
+    // generate unbounded metrics otherwise and OOM during fuzzing.
+    let peer_id = PeerId::ZERO;
+    // However, we want to choose a random _remote_ peer id to ensure we _don't_
+    // have metrics logging the remote peer id (which would eventually OOM in
+    // production for public-facing nodes).
+    let remote_peer_id = PeerId::random();
+
+    // Mock data
+    let network_context = NetworkContext::mock_with_peer_id(peer_id);
+    let socket = ReadOnlyTestSocketVec::new(data.to_vec());
+    let metadata = ConnectionMetadata::new(
+        remote_peer_id,
+        ConnectionId::from(123),
+        NetworkAddress::mock(),
+        ConnectionOrigin::Inbound,
+        MessagingProtocolVersion::V1,
+        SupportedProtocols::from(
+            [
+                ProtocolId::ConsensusRpc,
+                ProtocolId::ConsensusDirectSend,
+                ProtocolId::MempoolDirectSend,
+                ProtocolId::StateSynchronizerDirectSend,
+                ProtocolId::DiscoveryDirectSend,
+                ProtocolId::HealthCheckerRpc,
```

Comment:
sucks that we can't just add all enums easily, looks like we'd have to use a macro for that (https://docs.rs/strum_macros/0.19.4/strum_macros/derive.EnumIter.html)